core/domain/summary_services.py

```diff
@@ -352,29 +355,68 @@ def _generate_query(categories):
-
-    The return value is sorted in decreasing order of search rank.
+def require_activity_ids_to_be_public(activity_ids):
+    """Raises an exception if any activity id in the list does not exist, or
+    is not public, or if the list has duplicates.
     """
-    filtered_exp_summaries = [
-        exp_summary for exp_summary in
-        exp_services.get_featured_exploration_summaries().values()
-        if exp_summary.language_code in language_codes]
+    exploration_ids, collection_ids = utils.split_activity_ids_by_type(
+        activity_ids)
+
+    exp_summaries = exp_services.get_exploration_summaries_matching_ids(
+        exploration_ids)
+    col_summaries = collection_services.get_collection_summaries_matching_ids(
+        collection_ids)
+
+    for index, summary in enumerate(exp_summaries):
+        if summary is None:
+            raise Exception(
+                'Cannot feature non-existent exploration with id %s' %
+                exploration_ids[index])
+        if summary.status == rights_manager.ACTIVITY_STATUS_PRIVATE:
+            raise Exception(
+                'Cannot feature private exploration with id %s' % summary.id)
+
+    for index, summary in enumerate(col_summaries):
+        if summary is None:
+            raise Exception(
+                'Cannot feature non-existent collection with id %s' %
+                collection_ids[index])
+        if summary.status == rights_manager.ACTIVITY_STATUS_PRIVATE:
+            raise Exception(
+                'Cannot feature private collection with id %s' % summary.id)
+
+
+def get_featured_activity_summary_dicts(language_codes):
+    """Returns a list of featured activities with the given language code.
+
+    The return value is sorted according to the list stored in the datastore.
+    """
+    featured_activity_ids = library_services.get_featured_activity_ids()
 
-    search_ranks = {
-        exp_summary.id: exp_services.get_search_rank_from_exp_summary(
-            exp_summary)
-        for exp_summary in filtered_exp_summaries
-    }
+    exploration_ids_to_fetch, collection_ids_to_fetch = (
+        utils.split_activity_ids_by_type(featured_activity_ids))
 
-    sorted_exp_summaries = sorted(
-        filtered_exp_summaries,
-        key=lambda exp_summary: search_ranks[exp_summary.id],
-        reverse=True)
-
-    return _get_displayable_exp_summary_dicts(
-        sorted_exp_summaries, include_contributors=False)
+    exp_summary_dicts = get_displayable_exp_summary_dicts_matching_ids(
+        exploration_ids_to_fetch)
+    col_summary_dicts = get_displayable_collection_summary_dicts_matching_ids(
+        collection_ids_to_fetch)
+
+    activity_ids_to_summary_dicts = {}
+    for summary_dict in exp_summary_dicts:
+        activity_id = utils.get_activity_id(
+            feconf.ACTIVITY_TYPE_EXPLORATION, summary_dict['id'])
+        activity_ids_to_summary_dicts[activity_id] = summary_dict
+    for summary_dict in col_summary_dicts:
```

Comment:
I might be missing something, but what's your suggestion? I'm reluctant to add a separate function, since this seems like a very small thing -- in fact it's a single statement; I just broke it out into two because the invocation to construct activity_id felt a bit long.